storage/remote/queue_manager.go

```diff
@@ -551,11 +631,72 @@ outer:
 	return true
 }
 
-func (t *QueueManager) AppendExemplars(exemplars []record.RefExemplar) bool {
+func (t *QueueManager) updatePendingData(dataCount, dataSegment int) {
+	t.pendingDataMut.Lock()
+	defer t.pendingDataMut.Unlock()
+
+	t.pendingDataSegments[dataSegment] += dataCount
+
+	// We want to save segments whose data has been fully processed by the
+	// QueueManager. To avoid too much overhead when appending data, we'll only
+	// examine pending data per segment whenever a new segment is detected.
+	//
+	// This could cause our saved segment to lag behind multiple segments
+	// depending on the rate that new segments are created and how long
+	// data in other segments takes to be processed.
+	if dataSegment < t.lastMarkedSegment {
+		return
+	}
+
+	// We've received data for a new segment. We'll use this as a signal to see
+	// if older segments have been fully processed.
+	//
+	// We want to mark the highest segment which has no more pending data and is
+	// newer than our last mark.
+	t.latestDataSegment = dataSegment
+
+	var markableSegment int
+	for segment, count := range t.pendingDataSegments {
+		if segment >= dataSegment || count > 0 {
+			continue
+		}
+
+		if segment > markableSegment {
+			markableSegment = segment
+		}
+
+		// Clean up the pending map: the current segment has been completely
+		// consumed and doesn't need to be considered for marking again.
+		delete(t.pendingDataSegments, segment)
+	}
+
+	if markableSegment > t.lastMarkedSegment {
+		t.markSegment(markableSegment)
+		t.lastMarkedSegment = markableSegment
+	}
+}
+
+// processConsumedData uses data to update how many samples are still pending.
+// data is a map of segment index to consumed data count.
+func (t *QueueManager) processConsumedData(data map[int]int) {
+	t.pendingDataMut.Lock()
+	defer t.pendingDataMut.Unlock()
+
+	for segment, count := range data {
+		if _, tracked := t.pendingDataSegments[segment]; !tracked {
```

Comment:
we should probably include some debug logging and/or a metric, since we shouldn't expect to be attempting to subtract samples from `pendingDataSegments` for a segment we're no longer tracking